x-pack/plugins/alerting/server/lib/get_execution_log_aggregation.ts

```diff
@@ -0,0 +1,337 @@
+  timeoutMessage: estypes.AggregationsMultiBucketBase;
+  ruleExecution: {
+    executeStartTime: estypes.AggregationsMinAggregate;
+    executionDuration: estypes.AggregationsMaxAggregate;
+    scheduleDelay: estypes.AggregationsMaxAggregate;
+    esSearchDuration: estypes.AggregationsMaxAggregate;
+    totalSearchDuration: estypes.AggregationsMaxAggregate;
+    numTriggeredActions: estypes.AggregationsMaxAggregate;
+    outcomeAndMessage: estypes.AggregationsTopHitsAggregate;
+  };
+  alertCounts: IAlertCounts;
+  actionExecution: {
+    actionOutcomes: IActionExecution;
+  };
+}
+
+interface ExecutionUuidAggResult<TBucket = IExecutionUuidAggBucket>
+  extends estypes.AggregationsAggregateBase {
+  buckets: TBucket[];
+}
+export interface IExecutionLogAggOptions {
+  numExecutions: number;
+  page: number;
+  perPage: number;
+  sort: estypes.Sort;
+}
+
+const ExecutionLogSortFields: Record<string, string> = {
+  timestamp: 'ruleExecution>executeStartTime',
+  execution_duration: 'ruleExecution>executionDuration',
+  total_search_duration: 'ruleExecution>totalSearchDuration',
+  es_search_duration: 'ruleExecution>esSearchDuration',
+  schedule_delay: 'ruleExecution>scheduleDelay',
+  num_triggered_actions: 'ruleExecution>numTriggeredActions',
+};
+
+export function getExecutionLogAggregation({
+  numExecutions,
+  page,
+  perPage,
+  sort,
+}: IExecutionLogAggOptions) {
+  // Check if valid sort fields
+  const sortFields = flatMap(sort as estypes.SortCombinations[], (s) => Object.keys(s));
+  for (const field of sortFields) {
+    if (!Object.keys(ExecutionLogSortFields).includes(field)) {
+      throw Boom.badRequest(
+        `Invalid sort field "${field}" - must be one of [${Object.keys(ExecutionLogSortFields).join(
+          ','
+        )}]`
+      );
+    }
+  }
+
+  // Check if valid page value
+  if (page <= 0) {
+    throw Boom.badRequest(`Invalid page field "${page}" - must be greater than 0`);
+  }
+
+  // Check if valid page value
+  if (perPage <= 0) {
+    throw Boom.badRequest(`Invalid perPage field "${perPage}" - must be greater than 0`);
+  }
+
+  // Check if valid num executions
+  if (numExecutions > DEFAULT_MAX_BUCKETS_LIMIT) {
+    throw Boom.badRequest(
+      `Invalid numExecutions requested "${numExecutions}" - must be less than ${DEFAULT_MAX_BUCKETS_LIMIT}`
+    );
+  }
+
+  return {
+    // Get total number of executions
+    executionUuidCardinality: {
+      cardinality: {
+        field: EXECUTION_UUID_FIELD,
+      },
+    },
+    executionUuid: {
+      // Bucket by execution UUID
+      terms: {
+        field: EXECUTION_UUID_FIELD,
+        size: numExecutions,
+        order: formatSortForTermSort(sort),
+      },
+      aggs: {
+        // Bucket sort to allow paging through executions
+        executionUuidSorted: {
+          bucket_sort: {
+            sort: formatSortForBucketSort(sort),
+            from: (page - 1) * perPage,
+            size: perPage,
```

Comment:
Looks like we're going to need `gap_policy: 'insert_zeros'` to ensure we're not losing docs when sorting on fields that may be empty like `es_search_duration` or `total_search_duration_ms`. Details in [this other thread](https://github.com/elastic/kibana/pull/127339/files#r825240516) below. I'm working through adding some tests around sorting and just ran into this on my end.

Reply:
That makes sense! Added here: https://github.com/elastic/kibana/pull/127339/commits/8ca934de33f91af9b181b34fac5f053313c5d471